rfcs/0012-compaction-state-persistence.md

```diff
@@ -0,0 +1,800 @@
+
+    - Check If the count of runnning compactions is less than the threshold. If yes, continue
+
+    - Check if the source L0 SSTs and SRs are not part of any other compaction. If yes, continue.
+
+    - Check if the destination SR is not part of any other compaction. If yes, continue.
+
+    - Check if creating new SR(as part of L0 compaction) would it form a tiered SR group and get compacted. If yes, do not create the SR.
+
+    - Add compaction validations to verify correct group of sources and destinations are selected. Reference: https://github.com/slatedb/slatedb/blob/main/rfcs/0002-compaction.md#compactions.
+
+    - The existing validations in `submit_compaction` method.
+
+  Note: Invalid compactions would be dropped from the list of compactions during validation.
+
+4. When a compaction successfully validates, the status of the compaction is updated as `submitted` in the compaction_state and added/updated in the `new_compactions` list.
+
+5. Try writing the compactor_state to the next sequential .compactor file.
+
+    If file exists,
+
+      - If latest .compactor compactor_epoch > current compactor epoch, die (fenced)
+
+      - If latest .compactor compactor_epoch == current compactor epoch, reconcile the compactor_state and pass `new_compactions` object to Step(3). This process would continue until successful .compactor file write or the `new_compaction` object is empty.
+      [This would happen when the compactor has written a .compactor file]
+
+      - If latest .compactor compactor_epoch < current compactor epoch, panic
+      ( Compactor_epoch going backwards)
+
+    [When this step is successful, compaction is persisted in the .compactor file]
+
+Note: The validations added in this protocol are best effort as the authority to validate a compaction lies with the compactor daemon thread. For more details refer(https://github.com/slatedb/slatedb/pull/695#discussion_r2289989866)
+
+### Resuming Partial Compactions
+
+1. When the output SSTs(part of the partially completed destination SR) are fetched, pick the lastEntry(the lastEntry in lexicographic order) from the last SST of the SR. Possible Approaches:
+    - Add a lastKey in the metadata block of SST as suggested here:https://github.com/slatedb/slatedb/pull/695/files#r2243447106 similar to first key and fetch it from the metadata block 
+
+    - Once on the relevant SST, go to the last block by iterating the indexes. Iterate to the lastKey of the last block of the SST.[Agreed]
+
+2. Ignore completed L0 SSTs and move the iterator on each SR to a key >= lastKey on SST partition
+
+3. This is done by doing a binary search on a SR to find the right SST partition and then iterating the blocks of the SST till we find the Entry. 
+    [Note: A corner case: With monotonically overlapping SST ranges(specifically the last key), a key might be present across a contiguous range of SST in a SR]
```

Comment:
```suggestion
    (Note: There's a corner case where there might be monotonically overlapping SST ranges. Specifically, the last key might be present across a contiguous range of SST in a SR.)
```

@flaneur2020 I seem to recall you making an update to ensure this doesn't happen. Did that get merged in? I can't find it (or perhaps it already worked this way?).